### PR TITLE
Backchannel logout support

### PR DESCRIPTION
**Pull request by skion**

### CODE REVIEW

File: lib/omniauth/strategies/oidc/back_channel_logout_request.rb [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+
+        def back_channel_logout_request
+          @logout_request || begin
+            decode_logout_token(@request.params['logout_token'])
+            @request
+          end
+        end
+
+        def decode_logout_token(logout_token)
+            logout_jwt = JWT.decode logout_token, @options.client_secret, true, {:algorithm => 'HS256'}
```
Comment (skion) [inline lib/omniauth/strategies/oidc/back_channel_logout_request.rb:43]: We need to verify the issuer and the audience claims:

    { :iss => @options.issuer, :verify_iss => true, :aud => @options.client_id, :verify_aud => true }

Until this module support discovery the `iss` claim should be made a config option too, so a user can change it, like: `@options.issuer` default `https://connect.authentiq.io/`.

Also check for the required `jti` and `iat`:

    { :verify_iat => true, :verify_jti => true }

and I would suggest:

    { :leeway => 60 }

Comment (skion) [reply lib/omniauth/strategies/oidc/back_channel_logout_request.rb:43]: There can be only one algorithm indeed. In that case let's not hard code it, but extract it instead as a config option (name `@options.client_signed_response_alg` default `HS256`) so it can be changed and matched to a registered client.

File: lib/omniauth/strategies/oidc/back_channel_logout_request.rb [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+            end
+          ensure
+            return unless result
+            return result.finish
+          end
+        end
+
+        private
+
+        def back_channel_logout_request
```
Comment (skion) [inline lib/omniauth/strategies/oidc/back_channel_logout_request.rb:35]: `backchannel_logout_request` ?

File: lib/omniauth/strategies/authentiq.rb [hunk 3]
```diff
@@ -17,7 +19,7 @@ class Authentiq < OmniAuth::Strategies::OAuth2
       }
 
       # Get options from parameters
-      option :authorize_options, [:scope, :display]
+      option :authorize_options, [:scope]
```
Comment (skion) [inline lib/omniauth/strategies/authentiq.rb:22]: Is this change intentional in this PR?

File: lib/omniauth/strategies/authentiq.rb [hunk 4]
```diff
@@ -3,6 +3,8 @@
 module OmniAuth
   module Strategies
     class Authentiq < OmniAuth::Strategies::OAuth2
+      autoload :BackChannelLogoutRequest, 'omniauth/strategies/oidc/back_channel_logout_request'
```
Comment (skion) [inline lib/omniauth/strategies/authentiq.rb:6]: `backchannel_logout_request`?

File: lib/omniauth/strategies/oidc/back_channel_logout_request.rb [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+          @options = options
+
+          begin
+            # for
+            # sign_out_callback.call(*back_channel_logout_request)
+            # to actually happen
+            # a proc must be set in the devise.rb initializer of the rails app
+            result = sign_out_callback.call(*back_channel_logout_request)
+          rescue
+            return backchannel_logout_response(400, ['Bad request']).finish
```
Comment (skion) [inline lib/omniauth/strategies/oidc/back_channel_logout_request.rb:20]: I'd prefer to change the body content to the exception description if possible, so that also JWT exceptions (e.g. `raise JWT::VerificationError, 'Algorithm not supported'`) come through and are presented to user.

The "Bad request" body is redundant anyway since that is already in the HTTP status line.

File: lib/omniauth/strategies/oidc/back_channel_logout_request.rb [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+          else
+            if result
+              result = backchannel_logout_response(200, ['OK'])
```
Comment (skion) [inline lib/omniauth/strategies/oidc/back_channel_logout_request.rb:23]: Body: "Logout succeeded" ?

File: lib/omniauth/strategies/oidc/back_channel_logout_request.rb [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+            else
+              result = backchannel_logout_response(501, ['Not implemented'])
```
Comment (skion) [inline lib/omniauth/strategies/oidc/back_channel_logout_request.rb:25]: Also here the exception value, or the reason the logout didn't succeed?

File: lib/omniauth/strategies/oidc/back_channel_logout_request.rb [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+
+        def validate_logout_token(logout_jwt)
+          valid = false
+          begin
+            if logout_jwt[0].key?('iss') &&
+                logout_jwt[0].key?('sub') &&
+                logout_jwt[0].key?('aud') &&
+                logout_jwt[0].key?('iat') &&
+                logout_jwt[0].key?('jti') &&
+                logout_jwt[0].key?('events') &&
```
Comment (skion) [inline lib/omniauth/strategies/oidc/back_channel_logout_request.rb:57]: `aud`, `iss`, `iat` and `jti` can be removed here then.

File: lib/omniauth/strategies/oidc/back_channel_logout_request.rb (hunk 1 above)
Comment (skion) [inline lib/omniauth/strategies/oidc/back_channel_logout_request.rb:43]: Unless you need the JWT header (i.e. `logout_jwt[1]`) I would already strip it off right here so that elsewhere you can just reference the payload as `logout_jwt` instead of `logout_jwt[0]`. Usually you will only need the payload and not the header, except maybe right before/after decoding.

File: lib/omniauth/strategies/oidc/back_channel_logout_request.rb [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+        def sign_out_callback
+          @options[:enable_remote_sign_out]
+        end
+
+        def backchannel_logout_response(code, body)
+          response = Rack::Response.new
+          response.status = code
+          response['Cache-Control'] = 'no-cache, no-store'
+          response['Pragma'] = 'no-cache'
+          response.body = body
```
Comment (skion) [inline lib/omniauth/strategies/oidc/back_channel_logout_request.rb:75]: This body (if not null) will need a matching `Content-Type` header. You could just add `Content-Type: text/plain; charset=utf-8` and use the body as-is, or wrap it in HTML or so if you prefer.

File: lib/omniauth/strategies/authentiq.rb (hunk 3 above)
Comment (alexkeramidas) [reply lib/omniauth/strategies/authentiq.rb:22]: Just cleanup until we decide that its useful
